NPE during report aggregation when multiple modules are specified using -pl arg

Hi,

Consider the following multi-module maven project.

- Parent
  - child1 
  - child2
    - GrandChild1
    - GrandChild2
  - child3 

On running `mvn -pl child1,child2/GrandChild1 clean scoverage:report -e`, the command fails with the following null pointer exception:
```bash
[INFO] Reactor Summary:
[INFO]
[INFO] child1 1.0-SNAPSHOT ................................ SUCCESS [  7.912 s]
[INFO] child3 1.0-SNAPSHOT ................................ FAILURE [  5.126 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 13.185 s
[INFO] Finished at: 2022-02-02T21:57:03-08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.scoverage:scoverage-maven-plugin:1.4.11:report (default-cli) on project child3: An error has occurred in SCoverage report generation: Report generation exception: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.scoverage:scoverage-maven-plugin:1.4.11:report (default-cli) on project child3: An error has occurred in SCoverage report generation: Report generation exception
.
.
.
Caused by: java.lang.NullPointerException
    at org.scoverage.plugin.SCoverageReportMojo.rebase (SCoverageReportMojo.java:560)
    at org.scoverage.plugin.SCoverageReportMojo.generateAggregatedReports (SCoverageReportMojo.java:506)
    at org.scoverage.plugin.SCoverageReportMojo.generate (SCoverageReportMojo.java:230)
    at org.scoverage.plugin.SCoverageReportMojo.execute (SCoverageReportMojo.java:385)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
```

This is because neither child1 nor child3 returns true for `module.isExecutionRoot()` and thus `topLevelModule` is always null.
https://github.com/scoverage/scoverage-maven-plugin/blob/163bbe9df7f907de744772aa9db29f700070aa6e/src/main/java/org/scoverage/plugin/SCoverageReportMojo.java#L440-L446

--- a/README.md
+++ b/README.md
@@ -158,7 +158,8 @@ or
 There is no separate mojo for aggregated reports, there is `aggregate` parameter.
 To additionally generate aggregated SCoverage report for root module, when generating regular reports,
 set `aggregate` parameter value to `true`.
-It works only in multimodule projects.
+It works only in multimodule projects, the aggregated report will be generated in the current
+execution root.
 
 It can be configured by defining `aggregate` plugin configuration parameter or `scoverage.aggregate` project property.
 

--- a/src/it/test_report_for_some_submodules/invoker.properties
+++ b/src/it/test_report_for_some_submodules/invoker.properties
@@ -1,0 +1,1 @@
+invoker.goals=clean scoverage:report -e -ntp -pl module01 -pl module02/submodule02_02

--- a/src/it/test_report_for_some_submodules/module01/pom.xml
+++ b/src/it/test_report_for_some_submodules/module01/pom.xml
@@ -1,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <parent>
+        <groupId>it.scoverage-maven-plugin</groupId>
+        <artifactId>test_report_for_some_submodules</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>module01</artifactId>
+    <name>Test Scoverage report works for specified submodules: Module 1</name>
+
+</project>

--- a/src/it/test_report_for_some_submodules/module01/src/main/scala/pkg01/HelloService1.scala
+++ b/src/it/test_report_for_some_submodules/module01/src/main/scala/pkg01/HelloService1.scala
@@ -1,0 +1,12 @@
+package pkg01
+
+class HelloService1
+{
+  def hello =
+  {
+    "Hello from module 1"
+  }
+
+}
+
+object HelloService1 extends HelloService1

--- a/src/it/test_report_for_some_submodules/module01/src/test/scala/pkg01/HelloServiceTest.scala
+++ b/src/it/test_report_for_some_submodules/module01/src/test/scala/pkg01/HelloServiceTest.scala
@@ -1,0 +1,14 @@
+package pkg01
+
+import org.junit.Test;
+import org.junit.Assert.assertEquals
+
+class HelloServiceTest
+{
+    @Test
+    def test1()
+    {
+        assertEquals("Hello from module 1", HelloService1.hello)
+    }
+
+}

--- a/src/it/test_report_for_some_submodules/module02/pom.xml
+++ b/src/it/test_report_for_some_submodules/module02/pom.xml
@@ -1,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <parent>
+        <groupId>it.scoverage-maven-plugin</groupId>
+        <artifactId>test_report_for_some_submodules</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>module02</artifactId>
+    <packaging>pom</packaging>
+    <name>Test Scoverage report works for specified submodules: Module 2</name>
+
+    <modules>
+        <module>submodule02_01</module>
+        <module>submodule02_02</module>
+    </modules>
+
+
+</project>

--- a/src/it/test_report_for_some_submodules/module02/submodule02_01/pom.xml
+++ b/src/it/test_report_for_some_submodules/module02/submodule02_01/pom.xml
@@ -1,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <parent>
+        <groupId>it.scoverage-maven-plugin</groupId>
+        <artifactId>module02</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>submodule02_01</artifactId>
+    <name>Test Scoverage report works for specified submodules: SubModule 2_1</name>
+
+</project>

--- a/src/it/test_report_for_some_submodules/module02/submodule02_01/src/main/scala/pkg02_01/HelloService2.scala
+++ b/src/it/test_report_for_some_submodules/module02/submodule02_01/src/main/scala/pkg02_01/HelloService2.scala
@@ -1,0 +1,12 @@
+package pkg02_01
+
+class HelloService2
+{
+  def hello =
+  {
+    "Hello from submodule02_01"
+  }
+
+}
+
+object HelloService2 extends HelloService2

--- a/src/it/test_report_for_some_submodules/module02/submodule02_01/src/test/scala/pkg02_01/HelloServiceTest.scala
+++ b/src/it/test_report_for_some_submodules/module02/submodule02_01/src/test/scala/pkg02_01/HelloServiceTest.scala
@@ -1,0 +1,14 @@
+package pkg02_01
+
+import org.junit.Test;
+import org.junit.Assert.assertEquals
+
+class HelloServiceTest
+{
+    @Test
+    def test2()
+    {
+        assertEquals("Hello from submodule02_01", HelloService2.hello)
+    }
+
+}

--- a/src/it/test_report_for_some_submodules/module02/submodule02_02/pom.xml
+++ b/src/it/test_report_for_some_submodules/module02/submodule02_02/pom.xml
@@ -1,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <parent>
+        <groupId>it.scoverage-maven-plugin</groupId>
+        <artifactId>module02</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>submodule02_02</artifactId>
+    <name>Test Scoverage report works for specified submodules: SubModule 2_2</name>
+
+</project>

--- a/src/it/test_report_for_some_submodules/module02/submodule02_02/src/main/scala/pkg02_02/HelloService2.scala
+++ b/src/it/test_report_for_some_submodules/module02/submodule02_02/src/main/scala/pkg02_02/HelloService2.scala
@@ -1,0 +1,12 @@
+package pkg02_02
+
+class HelloService2
+{
+  def hello =
+  {
+    "Hello from submodule02_02"
+  }
+
+}
+
+object HelloService2 extends HelloService2

--- a/src/it/test_report_for_some_submodules/module02/submodule02_02/src/test/scala/pkg02_02/HelloServiceTest.scala
+++ b/src/it/test_report_for_some_submodules/module02/submodule02_02/src/test/scala/pkg02_02/HelloServiceTest.scala
@@ -1,0 +1,14 @@
+package pkg02_02
+
+import org.junit.Test;
+import org.junit.Assert.assertEquals
+
+class HelloServiceTest
+{
+    @Test
+    def test2()
+    {
+        assertEquals("Hello from submodule02_02", HelloService2.hello)
+    }
+
+}

--- a/src/it/test_report_for_some_submodules/pom.xml
+++ b/src/it/test_report_for_some_submodules/pom.xml
@@ -1,0 +1,65 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>it.scoverage-maven-plugin</groupId>
+        <artifactId>integration_tests_parent</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../integration_tests_parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>test_report_for_some_submodules</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+    <name>Test Scoverage report works for specified submodules</name>
+    <description>Test Scoverage report works for specified submodules</description>
+
+    <properties>
+        <scala.compat.version>2.13</scala.compat.version>
+        <scala.minor.version>12</scala.minor.version>
+    </properties>
+
+    <modules>
+        <module>module01</module>
+        <module>module02</module>
+    </modules>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>net.alchim31.maven</groupId>
+                <artifactId>scala-maven-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>@project.groupId@</groupId>
+                <artifactId>@project.artifactId@</artifactId>
+                <configuration>
+                    <aggregate>true</aggregate> <!-- for aggregated report -->
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <reporting>
+        <plugins>
+            <plugin>
+                <groupId>@project.groupId@</groupId>
+                <artifactId>@project.artifactId@</artifactId>
+                <configuration>
+                    <aggregate>true</aggregate> <!-- for aggregated report -->
+                </configuration>
+            </plugin>
+        </plugins>
+    </reporting>
+</project>

--- a/src/it/test_report_for_some_submodules/validate.groovy
+++ b/src/it/test_report_for_some_submodules/validate.groovy
@@ -1,0 +1,14 @@
+try {
+
+    def aggregatedScoverageFile = new File(basedir, "target/scoverage.xml")
+    assert aggregatedScoverageFile.exists()
+
+    def aggregatedReportFile = new File(basedir, "target/site/scoverage/index.html")
+    assert aggregatedReportFile.exists()
+
+    return true
+
+} catch (Throwable e) {
+    e.printStackTrace()
+    return false
+}

--- a/src/main/java/org/scoverage/plugin/SCoverageReportMojo.java
+++ b/src/main/java/org/scoverage/plugin/SCoverageReportMojo.java
@@ -23,9 +23,11 @@ import java.io.IOException;
 import java.io.File;
 import java.io.FileInputStream;
 import java.util.ArrayList;
+import java.util.ArrayDeque;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Locale;
+import java.util.Queue;
 import java.util.ResourceBundle;
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -533,6 +535,31 @@ public class SCoverageReportMojo
         {
             getLog().info( String.format( "Found %d subproject scoverage data directories.",
                     scoverageDataDirs.size() ) );
+        }
+
+        /* traverse up the module tree until a module isExecutionRoot */
+        if ( topLevelModule == null )
+        {
+            Queue<MavenProject> candidateForTopLevelModules = new ArrayDeque<>(reactorProjects);
+            while ( !candidateForTopLevelModules.isEmpty() )
+            {
+                MavenProject module = candidateForTopLevelModules.poll();
+                if ( module.isExecutionRoot() )
+                {
+                    topLevelModule = module;
+                    break;
+                }
+                if ( module.hasParent() )
+                {
+                    candidateForTopLevelModules.add(module.getParent());
+                }
+            }
+        }
+        if ( topLevelModule == null )
+        {
+            // This exception should never be thrown.
+            throw new IllegalStateException("Cannot find the top level module to write the " +
+                    "aggregated reports.");
         }
 
         File topLevelModuleOutputDirectory = rebase( outputDirectory, topLevelModule );

